File: src/main/java/gargoyle/ct/config/convert/impl/CTConfigDataConverter.java
package gargoyle.ct.config.convert.impl;

import gargoyle.ct.config.convert.CTUnitConverter;
import gargoyle.ct.util.CTTimeUtil;
import gargoyle.ct.util.Defend;
import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class CTConfigDataConverter implements CTUnitConverter<long[]> {
    private static final String COMMENTS = "#;'";
    private static final String MSG_CANNOT_PARSE_LINE_0 = "Cannot parse line: {0}";
    private static final String MSG_CANNOT_PARSE_LINE_0_INVALID_TIME_UNIT_1 =
            "Cannot parse line: {0}, invalid time " + "unit {1}";
    private static final String MSG_CANNOT_PARSE_LINE_INVALID_TIME_UNIT_0 =
            "Cannot parse line, invalid time unit " + "{0}";
    private static final String MSG_COMMENTED_LINE_0 = "Commented line: {0}";
    private static final String MSG_EMPTY_LINE = "Empty line";
    private static final String PATTERN_FORMAT = "{0}{1}/{2}{3}/{4}{5}";
    private static final Pattern PATTERN_PARSE = Pattern.compile(
            "(?:([0-9]+)([a-zA-Z]+))/(?:([0-9]+)([a-zA-Z]+))[^a-zA-Z0-9]+(?:([0-9]+)([a-zA-Z]+))",
            Pattern.CASE_INSENSITIVE);
    private static final String UNIT_HOURS = "H";
    private static final String UNIT_MINUTES = "M";
    private static final String UNIT_SECONDS = "S";

    @NotNull
    @Override
    public String format(@NotNull TimeUnit unit, long... data) {
        String unitChar;
        switch (unit) {
            case HOURS:
                unitChar = UNIT_HOURS;
                break;
            case MINUTES:
                unitChar = UNIT_MINUTES;
                break;
            case SECONDS:
                unitChar = UNIT_SECONDS;
                break;
            case DAYS:
            case MILLISECONDS:
            case MICROSECONDS:
            case NANOSECONDS:
            default:
                throw new UnsupportedOperationException(MessageFormat.format(MSG_CANNOT_PARSE_LINE_INVALID_TIME_UNIT_0,
                        unit.name()));
        }
        return MessageFormat.format(PATTERN_FORMAT,
                CTTimeUtil.fromMillis(unit, data[0]),
                unitChar,
                CTTimeUtil.fromMillis(unit, data[1]),
                unitChar,
                CTTimeUtil.fromMillis(unit, data[2]),
                unitChar);
    }

    @NotNull
    @Override
    public long[] parse(@NotNull String data) {
        Defend.notEmptyTrimmed(data, MSG_EMPTY_LINE);
        String trimmedLine = data.trim();
        Defend.isFalse(COMMENTS.contains(trimmedLine.substring(0, 1)), MessageFormat.format(MSG_COMMENTED_LINE_0,
                data));
        long[] longs = new long[3];
        Matcher m = PATTERN_PARSE.matcher(trimmedLine);
        if (m.find()) {
            int groupCount = m.groupCount();
            Defend.equals(groupCount, 6, MessageFormat.format(MSG_CANNOT_PARSE_LINE_0, data));
            for (int g = 1; g <= groupCount; g += 2) {
                String q = m.group(g);
                String u = m.group(g + 1);
                TimeUnit unit;
                switch (u) {
                    case UNIT_HOURS:
                        unit = TimeUnit.HOURS;
                        break;
                    case UNIT_MINUTES:
                        unit = TimeUnit.MINUTES;
                        break;
                    case UNIT_SECONDS:
                        unit = TimeUnit.SECONDS;
                        break;
                    default:
                        throw new IllegalArgumentException(MessageFormat.format(
                                MSG_CANNOT_PARSE_LINE_0_INVALID_TIME_UNIT_1,
                                data,
                                u));
                }
                try {
                    longs[g / 2] = CTTimeUtil.toMillis(unit, Long.parseLong(q));
                } catch (NumberFormatException ex) {
                    throw new IllegalArgumentException(MessageFormat.format(MSG_CANNOT_PARSE_LINE_0, data), ex);
                }
            }
        } else {
            throw new IllegalArgumentException(data);
        }
        return longs;
    }
}

File: app/src/main/kotlin/gargoyle/ct/config/convert/impl/CTConfigDataConverter.kt
package gargoyle.ct.config.convert.impl

import gargoyle.ct.util.util.CTTimeUtil
import kotlinx.serialization.KSerializer
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import java.text.MessageFormat
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

class CTConfigDataConverter(private val unit: TimeUnit = TimeUnit.MINUTES) : KSerializer<LongArray> {

    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("ConfigData", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): LongArray {
        val data = decoder.decodeString()
        require(data.isNotBlank()) { MSG_EMPTY_LINE }
        val trimmedLine = data.trim()
        require(!COMMENTS.contains(trimmedLine[0])) { ("Commented line: $data") }
        val longs = LongArray(3)
        val m = PATTERN_PARSE.matcher(trimmedLine)
        if (m.find()) {
            val groupCount = m.groupCount()
            require(groupCount == 6) { "Cannot parse line: $data" }
            var g = 1
            while (g <= groupCount) {
                val q = m.group(g)
                val unit: TimeUnit = when (val u = m.group(g + 1)) {
                    UNIT_HOURS -> TimeUnit.HOURS
                    UNIT_MINUTES -> TimeUnit.MINUTES
                    UNIT_SECONDS -> TimeUnit.SECONDS
                    else -> throw IllegalArgumentException("Cannot parse line: $data, invalid time unit $u")
                }
                try {
                    longs[g / 2] = CTTimeUtil.toMillis(unit, q.toLong())
                } catch (ex: NumberFormatException) {
                    throw IllegalArgumentException("Cannot parse line: $data", ex)
                }
                g += 2
            }
        } else {
            throw IllegalArgumentException(data)
        }
        return longs
    }


    override fun serialize(encoder: Encoder, value: LongArray) {
        val unitChar: String = when (unit) {
            TimeUnit.HOURS -> UNIT_HOURS
            TimeUnit.MINUTES -> UNIT_MINUTES
            TimeUnit.SECONDS -> UNIT_SECONDS
            TimeUnit.DAYS, TimeUnit.MILLISECONDS, TimeUnit.MICROSECONDS, TimeUnit.NANOSECONDS ->
                throw UnsupportedOperationException("Cannot parse line, invalid time unit ${TimeUnit.MINUTES.name}")

            else ->
                throw UnsupportedOperationException("Cannot parse line, invalid time unit ${TimeUnit.MINUTES.name}")
        }
        return encoder.encodeString(
            MessageFormat.format(
                PATTERN_FORMAT,
                CTTimeUtil.fromMillis(TimeUnit.MINUTES, value[0]),
                unitChar,
                CTTimeUtil.fromMillis(TimeUnit.MINUTES, value[1]),
                unitChar,
                CTTimeUtil.fromMillis(TimeUnit.MINUTES, value[2]),
                unitChar
            )
        )
    }

    companion object {
        private const val COMMENTS = "#;'"
        private const val MSG_EMPTY_LINE = "Empty line"
        private const val PATTERN_FORMAT = "{0}{1}/{2}{3}/{4}{5}"
        private val PATTERN_PARSE = Pattern.compile(
            "(?:([0-9]+)([a-zA-Z]+))/(?:([0-9]+)([a-zA-Z]+))[^a-zA-Z0-9]+(?:([0-9]+)([a-zA-Z]+))",
            Pattern.CASE_INSENSITIVE
        )
        private const val UNIT_HOURS = "H"
        private const val UNIT_MINUTES = "M"
        private const val UNIT_SECONDS = "S"
    }
}
